### PR TITLE
feat: upgrade ibc-go from v2 to v4.3.0 (with cosmos-sdk v0.45.12)

### PR DESCRIPTION
**Pull request by youngjoon-lee**
## Motivation

>Susannah, [Feb 2, 2023 22:12:17]:
Hello, I have queried medibloc and see that you are using v2.0.3 of ibc-go which is in the end of life. This means that patches and bug fixes are no longer back-ported to this version. We strongly encourage you to upgrade your version of ibc to v4.2.0, this is because v3.4.x will be in end of life in the middle of march. The migration documentation to go from v2 to v3 is here https://ibc.cosmos.network/main/migrations/v2-to-v3.html and for v3 to v4, take a look here https://ibc.cosmos.network/main/migrations/v3-to-v4.html 

## What I did
- Upgraded [ibc-go](https://github.com/cosmos/ibc-go) from v2 to v4
    - **Without** adopting the following new features:
        -  [ICS27](https://ibc.cosmos.network/main/migrations/v2-to-v3.html#ics27) Interchain Accounts (aka, ICA)
        - [ICS29](https://ibc.cosmos.network/main/migrations/v3-to-v4.html#ics29-fee-middleware) Fee Middleware. For more details, see a [comment](https://github.com/medibloc/panacea-core/pull/630#discussion_r1099812721) below.
    - This requires us to upgrade wasmd to v0.30.0+ and cosmos-sdk to v0.45.12+. For more details, please see a [comment](https://github.com/medibloc/panacea-core/pull/630#discussion_r1099777424) below.
    - There are a lot of changes in `app.go`. I put reference links as PR comments below.

## Tests
- Tested IBC transfers manually with a Panacea Localnet and the Osmosis Mainnet. 
    - Panacea -> Osmosis: [Success](https://www.mintscan.io/osmosis/txs/E02D3B85B867C6D8959590E9B913C5D263B539AEA31C43A39AC6B9C6F7BB2F05)
    - Osmosis -> Panacea: [Success](https://www.mintscan.io/osmosis/txs/469A6A65E25517CB915EB75F34456E1F264DE3E07B293007CBD4AE9713747F3A)    
- Also tested IBC transfers before and after upgrading the chain from v2.0.5 to HEAD.

## Future work
- Automate IBC tests by referring to e2e tests in [ibc-go](https://github.com/cosmos/ibc-go/tree/main/e2e), [gaia](https://github.com/cosmos/gaia/blob/main/tests/e2e/e2e_ibc_test.go), or [osmosis](https://github.com/osmosis-labs/osmosis/tree/main/tests/e2e).

### CODE REVIEW

File: app/ante.go [hunk 1]
```diff
@@ -62,7 +62,7 @@ func NewAnteHandler(options HandlerOptions) (sdk.AnteHandler, error) {
 		ante.NewSigGasConsumeDecorator(options.AccountKeeper, sigGasConsumer),
 		ante.NewSigVerificationDecorator(options.AccountKeeper, options.SignModeHandler),
 		ante.NewIncrementSequenceDecorator(options.AccountKeeper),
-		ibcante.NewAnteDecorator(options.IBCChannelkeeper),
+		ibcante.NewAnteDecorator(options.IBCKeeper),
```
Comment (youngjoon-lee) [inline app/ante.go:65]: https://ibc.cosmos.network/main/migrations/v2-to-v3.html#ante-decorator

File: app/app.go [hunk 2]
```diff
@@ -492,8 +492,8 @@ func New(
 
 	// Create static IBC router, add transfer route, then set and seal it
 	ibcRouter := porttypes.NewRouter()
-	ibcRouter.AddRoute(ibctransfertypes.ModuleName, transferModule)
-	ibcRouter.AddRoute(wasm.ModuleName, wasm.NewIBCHandler(app.wasmKeeper, app.IBCKeeper.ChannelKeeper))
+	ibcRouter.AddRoute(ibctransfertypes.ModuleName, transfer.NewIBCModule(app.TransferKeeper))
```
Comment (youngjoon-lee) [inline app/app.go:495]: https://ibc.cosmos.network/main/migrations/v2-to-v3.html#ibc-application-callbacks-moved-from-appmodule-to-ibcmodule

File: go.mod [hunk 3]
```diff
@@ -3,72 +3,75 @@ module github.com/medibloc/panacea-core/v2
 go 1.19
 
 require (
-	github.com/CosmWasm/wasmd v0.24.0
-	github.com/btcsuite/btcd v0.22.1
+	github.com/CosmWasm/wasmd v0.30.0
```
Comment (youngjoon-lee) [inline go.mod:6]: For IBC v4, `wasmd` must be also upgraded to v0.30.0 that is the first version based on IBC v4.
https://github.com/CosmWasm/wasmd/blob/main/CHANGELOG.md#v0300-2022-12-02

File: app/app.go [hunk 4]
```diff
@@ -408,10 +408,10 @@ func New(
 	// Create Transfer Keepers
 	app.TransferKeeper = ibctransferkeeper.NewKeeper(
 		appCodec, keys[ibctransfertypes.StoreKey], app.GetSubspace(ibctransfertypes.ModuleName),
+		app.IBCKeeper.ChannelKeeper, // may be replaced with middleware such as ics29 fee
 		app.IBCKeeper.ChannelKeeper, &app.IBCKeeper.PortKeeper,
 		app.AccountKeeper, app.BankKeeper, scopedTransferKeeper,
 	)
```
Comment (youngjoon-lee) [inline app/app.go:414]: From IBC v3, [IBC middleware](https://ibc.cosmos.network/main/ibc/middleware/develop.html) has been introduced.

As above, if we don't want to use any middleware for now, we can just pass the `app.IBCKeeper.ChannelKeeer` as a `ICS4Wrapper`. It's explained in the [IBC v2->v3 migration guide](https://ibc.cosmos.network/main/migrations/v2-to-v3.html#ics20). Also, I referenced the [ibc-go simapp](https://github.com/cosmos/ibc-go/commit/e40cdec6a#diff-d18972debee5e64f16e40807b2ae112ddbe609504a93ea5e1c80a5d489c3a08aR344).

Later, when we want to use any IBC middleware such as [IBC Fee middleware](https://ibc.cosmos.network/main/migrations/v3-to-v4.html#ics29-fee-middleware) like [this simapp](https://github.com/cosmos/ibc-go/pull/276/files#diff-d18972debee5e64f16e40807b2ae112ddbe609504a93ea5e1c80a5d489c3a08aR364).

File: app/app.go [hunk 5]
```diff
@@ -492,8 +492,8 @@ func New(
 
 	// Create static IBC router, add transfer route, then set and seal it
 	ibcRouter := porttypes.NewRouter()
-	ibcRouter.AddRoute(ibctransfertypes.ModuleName, transferModule)
-	ibcRouter.AddRoute(wasm.ModuleName, wasm.NewIBCHandler(app.wasmKeeper, app.IBCKeeper.ChannelKeeper))
+	ibcRouter.AddRoute(ibctransfertypes.ModuleName, transfer.NewIBCModule(app.TransferKeeper))
+	ibcRouter.AddRoute(wasm.ModuleName, wasm.NewIBCHandler(app.wasmKeeper, app.IBCKeeper.ChannelKeeper, app.IBCKeeper.ChannelKeeper))
```
Comment (youngjoon-lee) [inline app/app.go:496]: https://github.com/osmosis-labs/osmosis/pull/3838/files#diff-1ac782d2fab1d48ad7f7ef06cf2d58b37c99b6b4c98879a5ad86a985b73fad58R403

File: go.mod [hunk 6]
```diff
@@ -95,51 +97,51 @@ require (
...
 	gopkg.in/yaml.v3 v3.0.1 // indirect
 	nhooyr.io/websocket v1.8.6 // indirect
 )
 
 replace (
 	github.com/confio/ics23/go => github.com/cosmos/cosmos-sdk/ics23/go v0.8.0
-	github.com/cosmos/cosmos-sdk => github.com/medibloc/cosmos-sdk v0.45.9-panacea.1
-	github.com/cosmos/ledger-go => github.com/cosmos/ledger-go v0.9.3
+	// TODO: update this once a PR https://github.com/medibloc/cosmos-sdk/pull/182 is merged.
+	github.com/cosmos/cosmos-sdk => github.com/medibloc/cosmos-sdk v0.45.6-0.20230204134413-23923336f38f
```
Comment (youngjoon-lee) [inline go.mod:144]: // TODO: update this once a PR https://github.com/medibloc/cosmos-sdk/pull/182 is merged.

**Review by gyuguen — APPROVED** [review]
LGTM
Thank you for your hard working!

It seems to have applied well when I saw the migration documentation

**Review by H4NLee — APPROVED** [review]
lgtm.

File: go.mod (hunk 6 above)
Comment (youngjoon-lee) [reply go.mod]: 278089fd683d0ffc9b97efba10195a532869e675 Done

File: go.mod [hunk 7]
```diff
@@ -95,51 +96,50 @@ require (
...
+	golang.org/x/text v0.4.0 // indirect
+	google.golang.org/protobuf v1.28.2-0.20220831092852-f930b1dc76e8 // indirect
+	gopkg.in/ini.v1 v1.67.0 // indirect
 	gopkg.in/yaml.v2 v2.4.0 // indirect
 	gopkg.in/yaml.v3 v3.0.1 // indirect
 	nhooyr.io/websocket v1.8.6 // indirect
 )
 
 replace (
-	github.com/confio/ics23/go => github.com/cosmos/cosmos-sdk/ics23/go v0.8.0
```
Comment (youngjoon-lee) [inline go.mod:140]: This replace directive can be deleted as described in the release note of cosmos-sdk v0.45.12. https://github.com/cosmos/cosmos-sdk/releases/tag/v0.45.12

Comment (H4NLee) [reply go.mod:140]: we don't need this redirection anymore 👍

Comment (gyuguen) [reply go.mod:140]: nice

**Pull request merged**